vyper/exceptions.py

```
@@ -51,74 +51,117 @@ def __str__(self):
+    # altering list during iteration
+    # invalid use of range()
 
 
 class InvalidLiteralException(ParserException):
-    pass
+    """Invalid literal value."""
+    # number out of range
+    # too many decimals
+    # invalid hexstring length
+    # comparison against None
 
 
 class InvalidTypeException(ParserException):
-    pass
+    """Invalid type declaration."""
+    # unknown type
+    # unknown unit
+    # declaring a mapping inside a function
+    # len() on a type without length
+    # decimal exponentiation
+    # no type given
 
 
-class TypeMismatchException(ParserException):
-    pass
+class VariableDeclarationException(ParserException):
+    """Invalid variable declaration."""
+    # invalid name
+    # already declared / shadows existing value
+    # undeclared
 
 
 class FunctionDeclarationException(ParserException):
-    pass
+    """Invalid function declaration."""
+    # invalid arguments
+    # already declared / shadows existing value
 
 
 class EventDeclarationException(ParserException):
-    pass
+    """Invalid event declaration."""
+    # undeclared
+    # already declared / shadows existing value
+    # invalid argument types
+    # invalid indexed arg
 
 
-class VersionException(ParserException):
-    pass
+class TypeMismatchException(ParserException):
+    """TODO"""
+    # different types for operation
+    # same type but different units
+    # unexpected return value, expected return value
+    # left / right length mismatch
+    # list items mismatched
+    # type cannot be packed
 
 
-class SyntaxException(ParserException):
-    pass
+class ConstancyViolationException(ParserException):
+    """State-changing action inside a constant function."""
+
+
+class NonPayableViolationException(ParserException):
+    """Used msg.value in a nonpayable function."""
+
+
+class VersionException(ParserException):
+    """Version string is malform or incompatible with this version of Vyper."""
 
 
 class ArrayIndexException(ParserException):
```

`ArrayIndexException` is also only raised from one place.

---

This one seems fine, if you exceed the index of a static array it should throw a unique exception